Added Addon FrontierDistanceReachable

Shawn8901 opened the pull request:
This addon enables a check for frontier distance calculation. If this addon is enabled, a human path as to be within the triple distance length between both military buildings.

If the distance between both buildings is less then 10, the calculation base will a distance of 30 (maybe this value is a bit high?)

I hope that this "calculation" some how fits the request (#206).

@Flamefire: is there any test case for those frontier logic?

Flamefire commented:
I don't think there is a test case yet. You might write one by using a world with 2 players and placing a couple military buildings in varying distances. Then place some water between them and call the border functions again.

BTW: This won't work: If you have some stones or a building blocking the path to the other building then it would be considered "far". But the calculation is not retriggered when the stones are removed leaving that building exposed. You can use the free-path-finder directly and use a custom functor which only considers terrain instead of the human-path.

2 style issues: Use `std::max` instead of the `(x > y) ? x : y` pattern (here: `distance > 10`) as it is more expressive. And 2nd: When you are doing more work here, consider downloading llvm 6 or at least clang-format. CMake will pick it up from your path and add a clang-format target which you can use before committing (it only formats changed files, so it is fast). Otherwise try to stay close to the code-style around. This helps for avoiding merge conflicts and to many changes where only spaces were inserted/removed. What I have seen is the space after the `if`s

Shawn8901 commented:
Does the project use the default settings of clang-format?

Flamefire commented:
No. We have a custom clangformat file which is automatically picked up by clang-format with the `file` option (all encapsulated in the make target)
Our style is based on one of the clang-format styles and the most used style in our old codebase

Shawn8901 commented:
Thanks, i will give a try on that. About the issue with the pr, i will think about the resolution. And try to write a test case.

Shawn8901 commented:
Hi, 
sorry for the long time without update. I had/have trouble wit clang/LLVM and my SSD.... But i have now at least clang-format running. 
Just as question before i will work on a test case. 
do you mean a path finder like i have implemented in the last commit?

Flamefire commented:
First: Yes exactly. I think checking for water is enough. Just make sure, the water check actually checks for "unwalkable water" not just for water. If required change it to something that just requires walkable. Better might be to constrain the transition so that either left or right terrain must be walkable and neither must be dangerous.

Few issues: 

1. You need the `FreePathFinderImpl` in the cpp to instantiate the template correctly. Otherwise you'll get a linker error. (Your previous linker error was probably due to the anonymous namespace which excludes the symbol from the symbol table)
2. Then you don't need the `GetWorld` trait as you don't use it
3. Use World or GameWorldBase as the param for your checker class, no need for the template
4. Style: `namespace detail` is for something very specific to 1 task. Your pathfinder class does not fit there, as it might be used for something else, just like the human pathcheckers are used not solely for pathfinding but also for some direct node transition checks. A good way to use namespaces is to copy the directory structure (each directory=1 namespace, possibly nested). It is not done yet, except for small parts because when I started everything was in 1 folder.
5. Style: `FindXXX` as the name makes me assume it returns the found thing. Better use `HasXXX` or `[Does]XXXExists` or similar if all you return is a bool.
6. The pointer params to the path finders denote optional params. In your case there is no need for the road-vector as all you want to know is **IF** there is a way, not which.

Shawn8901 commented:
So i think i have now added everything requested.

Flamefire commented on src/addons/AddonFrontierDistanceReachable.h:
```diff
@@ -0,0 +1,41 @@
+#include "AddonBool.h"
+
+/**
+ *  Addon adds a 'order home' command button to military buildings. Pressing this button will send all soldiers of the highest rank
```
Change description or just remove

Flamefire commented on src/pathfinding/FindPathReachable.h:
```diff
@@ -0,0 +1,26 @@
+#ifndef FindPathReachable_h__
+#define FindPathReachable_h__
+
+#include "world/GameWorldBase.h"
```
Move include to cpp and simply fwd declare it here.

Shawn8901 replied:
i hope did mean what i did on the second commit.

Flamefire commented on src/pathfinding/PathConditionReachable.h:
```diff
@@ -0,0 +1,59 @@
+{
+    const GameWorldBase& world;
+
+	PathConditionReachable(const GameWorldBase& world) : world(world) {}
```
Please clang-format. Tabs and new lines are wrong. Maybe you have to rerun cmake or better add an empty new line to CMakeLists.txt to trigger it. (We are using `file(GLOB` so new files don't trigger a re-run like it should)

Shawn8901 replied:
Clang format just died before reaching this file. The reason was that i forgot to re-run cmake after a renaming.

Flamefire commented on src/test/testFrontierDistance.cpp:
```diff
@@ -0,0 +1,163 @@
+			break;
+	}
+	this->ggs.setSelection(AddonId::FRONTIER_DISTANCE_REACHABLE, 1);
+	for (int i = 1; i < 20; i++) {
```
better use `world.getHeight()` for clarity

Flamefire commented on src/test/testFrontierDistance.cpp:
```diff
@@ -0,0 +1,163 @@
+	}
+
+	this->ggs.setSelection(AddonId::FRONTIER_DISTANCE_REACHABLE,1 );
+	for (int i = 1; i < 20; i++) {
```
This is not `world.getHeight()` (50). Is that on purpose?

Shawn8901 replied:
no, just a copy & paste mistake.

Flamefire commented on src/test/testFrontierDistance.cpp:
```diff
@@ -0,0 +1,163 @@
+	}
+
+	this->ggs.setSelection(AddonId::FRONTIER_DISTANCE_REACHABLE, 0);
+	for (int i = 1; i < 20; i++) {
```
`world.getHeight()`

Flamefire commented on src/buildings/nobMilitary.cpp:
```diff
@@ -328,8 +329,14 @@ void nobMilitary::LookForEnemyBuildings(const nobBaseMilitary* const exception)
+            if(gwg->GetGGS().isEnabled(AddonId::FRONTIER_DISTANCE_REACHABLE)
+               && DoesReachablePathExist(*gwg, (*it)->GetPos(), pos, std::max(distance * 3, MILITARY_RADIUS[3])))
+            {
+                frontier_distance = 0;
```
Err, wait. `frontier_distance = 0` means "far". So I think it should have been `!DoesReachablePathExist`, or did I miss anything?

Shawn8901 replied:
i definitely with agree you!  Every testcase should have a partner with addon disabled. i dont know why i added a explicit "addoninactive". i will rename them, to match a naming scheme.
I will refactor those first. also i recognized that there is a test case doubled. 
I will take a look into it in the evening.

Flamefire commented on src/test/testFrontierDistance.cpp:
```diff
@@ -0,0 +1,180 @@
+    BOOST_REQUIRE_EQUAL(distance0, distance1);
+}
+
+BOOST_FIXTURE_TEST_CASE(FrontierDistanceMiddleAddonActive, FrontierWorldMiddle)
```
In the active test you use `FrontierWorldMiddle` but in the other you use `FrontierWorldSmall`. Maybe you could unify the 2 test cases to ensure this is the same? (Run in a loop, once with addon=1, once with addon=0 but keep the world setup the same. You could either issue a recalculation of the frontier distances or destroy and rebuild the buildings)

**Important**: Change the test case first before fixing the (possible) bug I mentioned above. IMO 1 test case MUST fail due to that bug. If not, either it is not a bug or you need another test case that triggers the bug.

Shawn8901 replied:
Okay i think i have managed it. there were multiple problems... The biggest was, that the water is set in my test case was walkable, so the whole test case was wrong... After i had fixed that i realized that the Condition check did not really  work for the test case.
I have fixed both problems and now i think it should work as expected.

Flamefire replied:
Now you are using unreachable terrain? Why not use the unwalkable water (like in some cases before)? That would be more common than lava. (or even test both)

What exactly was wrong with the condition check? IMO it was correct: You have to have all terrains around a point "not unreachable". Otherwise you could land/walk next to lava...

BTW: For your world structs you could use typedefs. They are shorter :)

Shawn8901 replied:
i have changed the test back to the "land part" of PathConditionHuman.
I have tried to unit test with water, but i have to be honest, i have no idea why its not working if i change it to water. for me it seems that water is threaded as walkable (maybe 1-2 fields??) because in real game scenario it works for me...)

Flamefire replied:
Found your problem. You put a single colon of water there. Just draw the triangles on a piece of paper and you will see why this fails. My blueprint (large so one can reuse it): 
![grid](https://user-images.githubusercontent.com/309017/35183504-f5b6c802-fde7-11e7-86d2-c6c3b695af88.png)

And with 1 column of water on t1 only (might be actually t2 but doesn't matter for showing): 
![water](https://user-images.githubusercontent.com/309017/35183591-13b3c282-fde9-11e7-9520-1012b682cb8b.jpg)


As you can easily see, walking is very much possible. You'd need to block it completely by making it at least 2 triangles, so one would have to walk ON water to cross. (Walking next to water is allowed)

Note: There is an edge case I just remembered: If there is a road, people can walk on water. A road is possible when every node has a) no danger terrain around and b) any terrain is (at least) walkable. This means that something like this should count as connected: 
![water2](https://user-images.githubusercontent.com/309017/35183593-19d1da5a-fde9-11e7-8a1d-a19f4a88cafa.jpg)
To handle this let the checker return true for any transition and only check nodes.

Shawn8901 commented:
Thanks for your explanation! It was very useful. 
Yesterday i was thinking about the idea to change the PathCondition. I rejected the idea (like you before), because we need PathConditionHuman and my implementation to be synchronized. Otherwise it could happen, that a building assume to be not attackable but can be reached by a soldier, which would result in a disaster.  
To ensure this, i changed PathConditionHuman to inherit from PathConditionReachable. The Check is copied from there and a soldier uses PathConditionHuman to evaluate its path. So everything should be fine now on my opinion.
Also i removed my magic value calculation for the distance. I found that there is a constant for the max distance a soldier will walk. So its better to use this.

For the unit test i have added more water, so every building should now be on an island. which is also an okay'ish test case. For the Near test case i also added a second test case to check if its also working for snow and lava.

Flamefire commented:
Almost perfect. I found another case I already mentioned. Try to rebuild the following as a test case: 
![rttrmap](https://user-images.githubusercontent.com/309017/35359140-3fc19d86-0159-11e8-976e-f4b00b6a946e.jpg)

All land except the dark blue (or even including the light blue but doesn't matter) which is water. I'm pretty sure your addon would classify the red military buildings as "far" because there is no path with the current pathfinder. However if one player builds a road over the water, then soldiers can walk over the water and attack the building which leads to the very case you were describing as a "disaster" ;-)

Build this test case and see if it fails. I think it can be fixed by moving the `IsEdgeOk` check to `PathConditionHuman` leaving the other one empty. That should cover everything although it might be to greedy in some cases (flags buildings as near that are separated by lava). To fix this the `IsEdgeOk` could be changed to simply check left&right for 'not unreachable' and omit the walkable check. This should completely fix this.

Shawn8901 commented:
Hi there, sorry for the delay.
I have prepared the test case (at least i think that its okay), also i have implemented your suggested fixes.
so i assume to build an island (5 fields water for map-edge) and 2 fields water as river on a map with size of 38 (middle).
I would now assume in the test case, that on both addon settings the frontier distance should be 1.

for this point the unit test is failing. i will try to look for that the next days but i wanted to share what i understand for the new test case.

Flamefire commented:
Hm, a simple rebase would have avoided the squash. But ok, found the changes.
First something that helps: As you check both addon settings in a loop it is hard to see which check fails. Idea: Add `i*10` to each distance. As the distance can only be 0-3 it is easy to tell apart (or use `i*100` etc.)
About the new test: This is not exactly right. Putting water at the border is ok, although not required. The map should be large enough, that it works. You could place the military buildings closer together (p0 on right and p1 on left of HQ) so that is the shorter distance. But the water at the border will also solve it for the addon case, so thats fine.
However the river in the middle is not correct. Let me try to visualize how your solution looks like:
![rttrmap](https://user-images.githubusercontent.com/309017/35591250-967f5f82-0609-11e8-9703-4cf7dafde4a2.png)
Ok, actually it does work. It is different from my idea but still works and is simpler. Good job!

So assuming the addon=1 test is the failing one (add the *100 trick) then the problem is that `PathConditionReachable::IsEdgeOk` is still there. That should go to solve that.
One minor issue: Your check for water in the test is `if(fieldDesc.kind == TerrainKind::WATER && (!fieldDesc.Is(ETerrain::Walkable) || fieldDesc.Is(ETerrain::Unreachable)))`, hence you allow unreachable water but actually you'd want unwalkable but reachable water. So: `if(fieldDesc.kind == TerrainKind::WATER && !fieldDesc.Is(ETerrain::Walkable) && !fieldDesc.Is(ETerrain::Unreachable))`
Otherwise you might get deadly water where no one can go near (-> "unreachable"). In the current config you'd find the regular water first, so you luckily don't run into this by now, but still should be change.

Shawn8901 commented:
Hi, i just tried a simple rebase before. but there was an conflict caused by the reformatting of GlobalSettings.cpp, where the Addons are registered. 
Maybe it would have been enough to solve the conflict on the first rebased commit, but i did not try this (to be honest).

The test case which fails is with addon enabled. it says that the military building is not reachable.
I tried to remove the IsEdgeOk-Code, but there is no difference, so i kept it, because i would prefer the fix you suggested earlier.

to be honest the point with the i*100 i do not understand.

With the commit i just changed the selection of water.

Flamefire commented:
Just for the future: Yes the resolving of the first commit would have been enough. Generally there would not have been less conflicts for squashing than for simple rebasing. But that's not really important now.

Ah yes I've just seen you changed IsEdgeOk to work as described.

I mean: Currently you get an error like "mismatch: 0 != 3". If you add `100*i` to each distance you would get "mismatch: 0 != 3" for i=0 and "mismatch: 100 != 103" for i=1 making it possible to know from which setting the error is.

I found your bug: You are setting water everywhere except at the border and the river. So exactly the opposite you wanted. Also you are starting at x/y =1 instead of 0 which is odd although not really wrong.
So basically set water instead of each continue or make all conditions together with an "or" each:
`if(curPoint.x < 5 || curPoint.x > world.GetWidth() - 5 || curPoint.y < 5 || curPoint.y > world.GetHeight() - 5 || (curPoint.x >= middle - 1 && curPoint.x <= middle)) set water;`

2 other things: Could you rename your "tUnwalkable" to "tWater"? IMO that would be easier to understand, what do you think?
And remotely related: Would you mind introducing an enum for the frontier distance instead of the numbers? I find that very hard to read what 0 or 3 means. Make it an enum in the class so you have to reference it like `nobMilitary::FAR` or so. Just make sure you don't change the meaning of the underlying ints (0=Far, 3=Nearest) to not break savegames

Shawn8901 commented:
God damit, i should just stay away from those loops....

about the enum: i fully agree with you and added it.

Flamefire commented:
Yeah I know the windows defines are annoying. You could prefix them with e.g. `DIST_`. I see you suffixed them which is less common but in this case ok. However I'd prefer to have all enum entries named similarly for consistency. Maybe `NEAR_HARBOR` is better? Just harbor seems ambigous:  `if(GetFrontierDistance() == HARBOR)` (what?) --> `if(GetFrontierDistance() == NEAR_HARBOR)` (reads better)

Last: You have `tUnreachable` in one place but check for unwalkable OR unreachable. I guess you don't want the unwalkable there (water is usually unwalkable but reachable, snow/lava is usually unreachable which includes unwalkable)

Flamefire commented:
Finally it works. Great!
I was thinking: maybe we should make this Addon the default? I'm not sure what s2 did for buildings that are separated by water but I think others said it was like this Addon. Could you test that? (maybe create a simple map in s2 with 2players and water in the middle then build a military bld and wait for the ai to do the same)

Shawn8901 commented:
On my opinion this was already tested in the issue related to the pr (#206). so w/o addon its working as in the original.

Flamefire commented:
Ok thanks a lot
